src/material/core/datetime/date-selection-model.ts

```diff
@@ -0,0 +1,149 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {DateAdapter} from '@angular/material/core';
+import {Subject} from 'rxjs';
+
+export abstract class MatDateSelectionModel<D> {
+  private _valueChangesSubject = new Subject<void>();
+  valueChanges = this._valueChangesSubject.asObservable();
+
+  constructor(protected readonly adapter: DateAdapter<D>) {}
+
+  dispose() {
+    this.valueChanges.complete();
+  }
+
+  abstract add(date: D | null): void;
+  abstract isComplete(): boolean;
+  abstract getFirstSelectedDate(): D|null;
+  abstract getLastSelectedDate(): D|null;
+  abstract isSame(other: MatDateSelectionModel<D>): boolean;
+  abstract isValid(): boolean;
+}
+
+export interface DateRange<D> {
+  start: D | null;
+  end: D | null;
+}
+
+/**
+ * Concrete implementation of a MatDateSelectionModel that holds a single date.
+ */
+export class MatSingleDateSelectionModel<D> extends MatDateSelectionModel<D> {
+  private date: D | null = null;
+
+  constructor(adapter: DateAdapter<D>, date?: D | null) {
+    super(adapter);
+    if (date) {
+      this.date = date;
+    }
+  }
+
+  add(date: D | null) {
+    this.date = date;
+    this.valueChanges.next();
+  }
+
+  compareDate(other: MatSingleDateSelectionModel<D>) {
+    const date = this.asDate();
+    const otherDate = other.asDate();
+    if (date && otherDate) {
```

Comment:
I meant to do the `===` only if one of the dates is `null` or `undefined`, if they're both real date objects then we want to use `compareDate`. Like this:

```ts
if (date != null && otherDate != null) {
  return this.adapter.compareDate(date, otherDate);
} 
return date === otherDate;
```